crates/red_knot_python_semantic/src/types.rs

```diff
@@ -722,6 +724,9 @@ impl<'db> Type<'db> {
     /// wrong `false` answers in some cases.
     pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
         match (self, other) {
+            (Type::TypeAlias(alias), other) | (other, Type::TypeAlias(alias)) => {
+                alias.value(db).is_disjoint_from(db, other)
+            }
```

Comment:
This is confusing the type represented by the type alias when we see it used in an annotation expression (that is, the type spelled by the type expression on the RHS of the alias), with the type of the alias object itself (which is just a singleton known instance of `typing.TypeAliasType`). So the "de-reference" to the alias' value that you are implementing here will happen when we encounter a type alias in `infer_type_expression`, we don't need to implement it throughout the type system.

Since a type alias object is a singleton instance, it's always disjoint -- but if we use `KnownInstance`, we'll get that for free.

(I won't comment on all the other places below where you're matching on `TypeAlias` and implementing this de-reference, but the same comment applies to all of them.)

Reply:
It might be good to add a few tests that would have failed based on this -- e.g. showing that we understand that two different type alias singleton types are disjoint and distinct types, even if they are an alias to the same type.